crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -655,4 +927,574 @@ mod tests {
+        }));
+
+        let index = ReceiptSelector::create_sync_index(events.iter().chain(&[ev6]));
+
+        assert_eq!(*index.get(event_id!("$1")).unwrap(), 0);
+        assert_eq!(*index.get(event_id!("$2")).unwrap(), 1);
+        assert_eq!(*index.get(event_id!("$3")).unwrap(), 2);
+        assert_eq!(*index.get(event_id!("$4")).unwrap(), 3);
+        assert_eq!(*index.get(event_id!("$5")).unwrap(), 4);
+        assert_eq!(index.get(event_id!("$6")), None);
+
+        assert_eq!(index.len(), 5);
+
+        // Sync order are set according to the position in the vector.
+        let index = ReceiptSelector::create_sync_index(
+            [events[1].clone(), events[2].clone(), events[4].clone()].iter(),
+        );
+
+        assert_eq!(*index.get(event_id!("$2")).unwrap(), 0);
+        assert_eq!(*index.get(event_id!("$3")).unwrap(), 1);
+        assert_eq!(*index.get(event_id!("$5")).unwrap(), 2);
+
+        assert_eq!(index.len(), 3);
+    }
+
+    #[test]
+    fn test_receipt_selector_try_select_better() {
+        let events = make_test_events(user_id!("@bob:example.org"));
+
+        {
+            // No initial active receipt, so the first receipt we get *will* win.
+            let mut selector = ReceiptSelector::new(&vec![].into(), None);
+            selector.try_select_better(event_id!("$1"), 0);
+            let best_receipt = selector.finish();
+            assert_eq!(best_receipt.unwrap().event_id, event_id!("$1"));
+        }
+
+        {
+            // $3 is at pos 2, $1 at position 0, so $3 wins => no new change.
+            let mut selector = ReceiptSelector::new(&events, Some(event_id!("$3")));
+            selector.try_select_better(event_id!("$1"), 0);
+            let best_receipt = selector.finish();
+            assert!(best_receipt.is_none());
+        }
+
+        {
+            // The initial active receipt is returned, when it's part of the scanned
+            // elements.
+            let mut selector = ReceiptSelector::new(&events, Some(event_id!("$1")));
+            selector.try_select_better(event_id!("$1"), 0);
+            let best_receipt = selector.finish();
+            assert_eq!(best_receipt.unwrap().event_id, event_id!("$1"));
+        }
+
+        {
+            // $3 is at pos 2, $4 at position 3, so $4 wins.
+            let mut selector = ReceiptSelector::new(&events, Some(event_id!("$3")));
+            selector.try_select_better(event_id!("$4"), 3);
+            let best_receipt = selector.finish();
+            assert_eq!(best_receipt.unwrap().event_id, event_id!("$4"));
+        }
+    }
+
+    #[test]
+    fn test_receipt_selector_handle_pending_receipts() {
+        let sender = user_id!("@bob:example.org");
+        let ev1 = sync_timeline_message(sender, event_id!("$1"), "yo");
+        let ev2 = sync_timeline_message(sender, event_id!("$2"), "well?");
+        let events: Vector<_> = vec![ev1, ev2].into();
+
+        // Each test must be duplicated here:
```

Comment:
I would personally prefer separate functions, with names that make it clear they are similar, or maybe even a submodule if that is allowed?

Reply:
I wasn't very excited about that, because that duplicates a bit of test setup, but it's ok, it's test code :-). I've usually grouped tests in pairs (in one case, in triplet), so there's still some logical structure here and a bit more information density in those test names :-)